metricbeat/module/kafka/consumergroup/query.go

```diff
@@ -38,51 +37,64 @@ func fetchGroupInfo(
 
 	debugf("known consumer groups: ", groups)
 
-	wg := sync.WaitGroup{}
-	results := make(chan result, len(groups))
-	for _, group := range groups {
-		group := group
+	assignments, err := fetchGroupAssignments(b, groups)
+	if err != nil {
+		logp.Err("failed to fetch kafka group assignments: %v", err)
+		return err
+	}
+	if len(assignments) == 0 {
+		return nil
+	}
 
-		wg.Add(1)
-		go func() {
-			defer wg.Done()
-			resp, err := fetchGroupOffset(b, group, topicsFilter)
-			if err != nil {
-				logp.Err("failed to fetch '%v' group offset: %v", group, err)
+	results := make(chan result)
+	waiting := 0
+	for group, topics := range assignments {
+		// generate the map topic to partitions
+		queryTopics := make(map[string][]int32)
+		for topic, partitions := range topics {
+			if topicsFilter != nil && !topicsFilter(topic) {
+				continue
 			}
-			results <- result{err, group, resp}
-		}()
-	}
 
-	go func() {
-		wg.Wait()
-		close(results)
-	}()
+			// copy partition ids
+			count := len(partitions)
+			if count == 0 {
+				continue
+			}
 
-	assignments, err := fetchGroupAssignments(b, groups)
-	if err != nil {
-		// wait for workers to stop and drop results
-		for range results {
+			ids, i := make([]int32, count), 0
+			for partition := range partitions {
+				ids[i], i = partition, i+1
+			}
+			queryTopics[topic] = ids
 		}
 
-		return err
+		if len(queryTopics) == 0 {
+			continue
+		}
+
+		// fetch group offset
+		waiting++
+		go func(group string, partitions map[string][]int32, assign map[string]map[int32]groupAssignment) {
+			resp, err := fetchGroupOffset(b, group, partitions)
+			if err != nil {
+				logp.Err("failed to fetch '%v' group offset: %v", group, err)
+			}
+			results <- result{err, group, assign, resp}
+		}(group, queryTopics, topics)
 	}
 
-	for ret := range results {
-		if err := ret.err; err != nil {
-			// wait for workers to stop and drop results
-			for range results {
+	for waiting > 0 {
+		ret := <-results
+		waiting--
+		if ret.err != nil {
+			if err == nil {
+				err = ret.err
 			}
-			return err
+			continue
```

Comment:
we can stop processing data after having seen an error by doing this:

```
for waiting > 0 {
  ret := <-results
  waiting--
  if ret.err != nil && err == nil {
    err = ret.err
  }
  if err != nil {
    continue
  }
}
```

This will keep the waiting loop alive, assign the first `ret.err` only, but also stop processing any further results (results are discarded anyways).

Reply:
@urso I fix it.